File: Luca/app/src/test/java/de/culture4life/luca/history/HistoryManagerTest.java
package de.culture4life.luca.history;

import androidx.test.runner.AndroidJUnit4;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.annotation.Config;

import java.util.UUID;
import java.util.concurrent.TimeUnit;

import de.culture4life.luca.LucaUnitTest;
import de.culture4life.luca.checkin.CheckInData;
import de.culture4life.luca.children.ChildrenManager;
import de.culture4life.luca.crypto.CryptoManager;
import de.culture4life.luca.network.NetworkManager;
import de.culture4life.luca.preference.PreferencesManager;
import de.culture4life.luca.registration.RegistrationManager;

@RunWith(AndroidJUnit4.class)
@Config(sdk = 28)
public class HistoryManagerTest extends LucaUnitTest {

    ChildrenManager childrenManager;
    HistoryManager historyManager;
    CheckInData checkInData = new CheckInData();

    @Before
    public void setUp() {
        PreferencesManager preferencesManager = new PreferencesManager();
        NetworkManager networkManager = new NetworkManager();
        CryptoManager cryptoManager = new CryptoManager(preferencesManager, networkManager);
        RegistrationManager registrationManager = new RegistrationManager(preferencesManager, networkManager, cryptoManager);
        childrenManager = new ChildrenManager(preferencesManager, registrationManager);
        historyManager = new HistoryManager(preferencesManager, childrenManager);
        historyManager.initialize(application).blockingAwait();

        checkInData.setTraceId("asdf");
        checkInData.setLocationId(UUID.randomUUID());
    }

    @Test
    public void getItems_initially_hasNoValues() {
        historyManager.getItems().test().assertNoValues();
    }

    @Test
    public void getItems_afterAddingCheckInItem_hasValue() {
        historyManager.addCheckInItem(checkInData)
                .andThen(historyManager.getItems())
                .test().assertValueCount(1);
    }

    @Test
    public void getItems_afterClearItems_hasNoValues() {
        historyManager.addCheckInItem(checkInData)
                .andThen(historyManager.clearItems())
                .andThen(historyManager.getItems())
                .map(historyItem -> historyItem.type)
                .test()
                .assertValue(HistoryItem.TYPE_DATA_DELETED)
                .assertValueCount(1);
    }

    @Test
    public void deleteOldItems_hasOldItems_removesOldItems() {
        checkInData.setTimestamp(System.currentTimeMillis() - TimeUnit.DAYS.toMillis(29));
        historyManager.addCheckInItem(checkInData)
                .andThen(historyManager.deleteOldItems())
                .andThen(historyManager.getItems())
                .test()
                .assertNoValues();
    }

    @Test
    public void deleteOldItems_noOldItems_keepsCurrentItems() {
        checkInData.setTimestamp(System.currentTimeMillis() - TimeUnit.DAYS.toMillis(23));
        historyManager.addCheckInItem(checkInData)
                .andThen(historyManager.deleteOldItems())
                .andThen(historyManager.getItems())
                .test().assertValueCount(1);
    }

}

File: Luca/app/src/test/java/de/culture4life/luca/history/HistoryManagerTest.kt
package de.culture4life.luca.history

import androidx.test.runner.AndroidJUnit4
import de.culture4life.luca.LucaUnitTest
import de.culture4life.luca.checkin.CheckInData
import de.culture4life.luca.children.ChildrenManager
import de.culture4life.luca.crypto.CryptoManager
import de.culture4life.luca.genuinity.GenuinityManager
import de.culture4life.luca.network.NetworkManager
import de.culture4life.luca.preference.PreferencesManager
import de.culture4life.luca.registration.RegistrationManager
import de.culture4life.luca.util.TimeUtil
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config
import java.util.*
import java.util.concurrent.TimeUnit

@RunWith(AndroidJUnit4::class)
@Config(sdk = [28])
class HistoryManagerTest : LucaUnitTest() {

    private val preferencesManager = PreferencesManager()
    private val networkManager = NetworkManager()
    private val genuinityManager = GenuinityManager(preferencesManager, networkManager)
    private val cryptoManager = CryptoManager(preferencesManager, networkManager, genuinityManager)
    private val registrationManager = RegistrationManager(preferencesManager, networkManager, cryptoManager)
    private val childrenManager = ChildrenManager(preferencesManager, registrationManager)
    private val historyManager = getInitializedManager(HistoryManager(preferencesManager, childrenManager))
    private val checkInData = CheckInData().apply {
        traceId = "asdf"
        locationId = UUID.randomUUID()
    }

    @Test
    fun items_initially_hasNoValues() {
        historyManager.items.test().assertNoValues()
    }

    @Test
    fun items_afterAddingCheckInItem_hasValue() {
        historyManager.addCheckInItem(checkInData)
            .andThen(historyManager.items)
            .test().assertValueCount(1)
    }

    @Test
    fun items_afterClearItems_hasNoValues() {
        historyManager.addCheckInItem(checkInData)
            .andThen(historyManager.clearItems())
            .andThen(historyManager.items)
            .map { historyItem -> historyItem.type }
            .test()
            .assertValue(HistoryItem.TYPE_DATA_DELETED)
            .assertValueCount(1)
    }

    @Test
    fun deleteOldItems_hasOldItems_removesOldItems() {
        checkInData.timestamp = TimeUtil.getCurrentMillis() - TimeUnit.DAYS.toMillis(29)
        historyManager.addCheckInItem(checkInData)
            .andThen(historyManager.deleteOldItems())
            .andThen(historyManager.items)
            .test()
            .assertNoValues()
    }

    @Test
    fun deleteOldItems_noOldItems_keepsCurrentItems() {
        checkInData.timestamp = TimeUtil.getCurrentMillis() - TimeUnit.DAYS.toMillis(23)
        historyManager.addCheckInItem(checkInData)
            .andThen(historyManager.deleteOldItems())
            .andThen(historyManager.items)
            .test().assertValueCount(1)
    }

    @Test
    fun createOrderedList_withItems_isCorrect() {
        val listString = HistoryManager.createOrderedList(listOf("A", "B"))
        assertEquals("1\t\t\tA\n2\t\t\tB", listString)
    }

    @Test
    fun createUnorderedList_withItems_isCorrect() {
        val listString = HistoryManager.createUnorderedList(listOf("A", "B"))
        assertEquals("- A\n- B", listString)
    }

}